switchbot/__init__.py

```diff
@@ -28,23 +28,79 @@
 PRESS_KEY_SUFFIX = "00"
 
 _LOGGER = logging.getLogger(__name__)
+CONNECT_LOCK = threading.Lock()
+
+
+def _process_wohand(data) -> dict:
+    """Process woHand/Bot services data."""
+    _bot_data = {}
+
+    _sensor_data = binascii.unhexlify(data.encode())
+
+    # 128 switch or 0 press.
+    _bot_data["switchMode"] = bool(_sensor_data[1] & 0b10000000)
+
+    # 64 off or 0 for on, if not inversed in app.
+    if _bot_data["switchMode"]:
+        _bot_data["isOn"] = not bool(_sensor_data[1] & 0b01000000)
+
+    else:
+        _bot_data["isOn"] = False
+
+    _bot_data["battery"] = _sensor_data[2] & 0b01111111
+
+    return _bot_data
+
+
+def _process_wocurtain(data, reverse=True) -> dict:
+    """Process woCurtain/Curtain services data."""
+    _curtain_data = {}
+
+    _sensor_data = binascii.unhexlify(data.encode())
+
+    _curtain_data["calibration"] = bool(_sensor_data[1] & 0b01000000)
+    _curtain_data["battery"] = _sensor_data[2] & 0b01111111
+    _position = max(min(_sensor_data[3] & 0b01111111, 100), 0)
+    _curtain_data["position"] = (100 - _position) if reverse else _position
+
+    # light sensor level (1-10)
+    _curtain_data["lightLevel"] = (_sensor_data[4] >> 4) & 0b00001111
+
+    return _curtain_data
+
+
+def _process_wosensorth(data) -> dict:
+    """Process woSensorTH/Temp sensor services data."""
+    _wosensorth_data = {}
+
+    _sensor_data = binascii.unhexlify(data.encode())
+
+    _temp_sign = _sensor_data[4] & 0b10000000
+    _temp_c = _temp_sign * ((_sensor_data[4] & 0b01111111) + (_sensor_data[3] / 10))
+    _temp_f = (_temp_c * 9 / 5) + 32
+    _temp_f = (_temp_f * 10) / 10
+
+    _wosensorth_data["temp"]["c"] = _temp_c
+    _wosensorth_data["temp"]["f"] = _temp_f
+
+    _wosensorth_data["fahrenheit"] = bool(_sensor_data[5] & 0b10000000)
+    _wosensorth_data["humidity"] = _sensor_data[5] & 0b01111111
+    _wosensorth_data["battery"] = _sensor_data[2] & 0b01111111
+
+    return _wosensorth_data
 
 
 class SwitchbotDevice:
-    # pylint: disable=too-few-public-methods
-    # pylint: disable=too-many-instance-attributes
     """Base Representation of a Switchbot Device."""
 
-    def __init__(self, mac, password=None, interface=None, **kwargs) -> None:
+    def __init__(self, mac=None, password=None, interface=None, **kwargs) -> None:
```

Comment:
When is `mac` not required?

Reply:
For the discovery function.

SwitchbotDevice().discover() uses btle advertisements and doesn't require a MAC address. It will return all device advertisements for devices in range. (With a single scan)

The uuid is used to identify switchbot devices.

Reply:
You never update the mac address later?
So `_connect()` and `_writekey()` will fail?

Reply:
"mac" is a requirement again.